src/builder.rs

```diff
@@ -780,45 +969,78 @@ fn build_bundle<B, R: RngCore>(
         return Err(BuildError::OutputsDisabled);
     }
 
-    let num_actions = bundle_type
-        .num_actions(num_requested_spends, num_requested_outputs)
-        .map_err(|_| BuildError::BundleTypeNotSatisfiable)?;
-
     // Pair up the spends and outputs, extending with dummy values as necessary.
     let (pre_actions, bundle_meta) = {
-        let mut indexed_spends = spends
-            .into_iter()
-            .chain(iter::repeat_with(|| SpendInfo::dummy(&mut rng)))
-            .enumerate()
-            .take(num_actions)
-            .collect::<Vec<_>>();
+        // Use Vec::with_capacity().extend(...) instead of .collect() to avoid reallocations,
+        // as we can estimate the vector size beforehand.
+        let mut indexed_spends_outputs =
+            Vec::with_capacity(spends.len().max(outputs.len()).max(MIN_ACTIONS));
 
-        let mut indexed_outputs = outputs
-            .into_iter()
-            .chain(iter::repeat_with(|| OutputInfo::dummy(&mut rng)))
-            .enumerate()
-            .take(num_actions)
-            .collect::<Vec<_>>();
+        let spends_outputs_by_asset = partition_by_asset(&spends, &outputs);
+
+        indexed_spends_outputs.extend(spends_outputs_by_asset.into_iter().flat_map(
+            |(asset, (spends, outputs))| {
+                let num_asset_pre_actions = spends.len().max(outputs.len());
+
+                let first_spend = spends.first().map(|(s, _)| s.clone());
+
+                let mut indexed_spends = spends
+                    .into_iter()
+                    .chain(iter::repeat_with(|| {
+                        (
+                            pad_spend(first_spend.as_ref(), asset, &mut rng)
+                                .unwrap_or_else(|err| panic!("{:?}", err)),
+                            None,
+                        )
+                    }))
+                    .take(num_asset_pre_actions)
+                    .collect::<Vec<_>>();
+
+                let mut indexed_outputs = outputs
+                    .into_iter()
+                    .chain(iter::repeat_with(|| {
+                        (OutputInfo::dummy(&mut rng, asset), None)
+                    }))
+                    .take(num_asset_pre_actions)
+                    .collect::<Vec<_>>();
 
-        // Shuffle the spends and outputs, so that learning the position of a
-        // specific spent note or output note doesn't reveal anything on its own about
-        // the meaning of that note in the transaction context.
-        indexed_spends.shuffle(&mut rng);
-        indexed_outputs.shuffle(&mut rng);
+                // Shuffle the spends and outputs, so that the position does not reveal any
+                // information about the content.
```

Comment:
_From str4d ([r2602532807](https://github.com/zcash/orchard/pull/471#discussion_r2602532807))_

"information about the content" is not what this shuffle is protecting (that is protected by the encryption). The previous comment was correct; preserve it.
```suggestion
                // Shuffle the spends and outputs, so that learning the position of a
                // specific spent note or output note doesn't reveal anything on its own
                // about the meaning of that note in the transaction context.
```

Reply:
FIXED